lib/SIL/MemAccessUtils.cpp

```diff
@@ -21,6 +21,56 @@
 
 using namespace swift;
 
+SILValue swift::stripAccessMarkers(SILValue V) {
+  while (true) {
+    switch (V->getKind()) {
+    default:
+      return V;
+
+    case ValueKind::BeginBorrowInst:
+    case ValueKind::BeginAccessInst:
+      V = cast<SingleValueInstruction>(V)->getOperand(0);
+    }
+  }
+}
+
+// TODO: When the optimizer stops stripping begin_access markers, then we should
+// be able to assert that the result is a BeginAccessInst and the default case
+// is unreachable.
+SILValue swift::getAccessedAddress(SILValue v) {
+  assert(v->getType().isAddress());
+  while (true) {
+    switch (v->getKind()) {
+    default:
+      return v;
+
+    case ValueKind::BeginBorrowInst:
+    case ValueKind::StructElementAddrInst:
+    case ValueKind::TupleElementAddrInst:
+    case ValueKind::UncheckedTakeEnumDataAddrInst:
+    case ValueKind::TailAddrInst:
+    case ValueKind::IndexAddrInst:
+      v = cast<SingleValueInstruction>(v)->getOperand(0);
+      continue;
+    };
+  }
+}
+
+bool swift::isLetAddress(SILValue accessedAddress) {
+  assert(accessedAddress == getAccessedAddress(accessedAddress)
```

Comment:
Why not just do this for the user rather than assert it?

Reply:
Because whoever uses this needs to be aware that they are using the correct address to represent a formal access, and use the same address for other comparisons and queries. We should not try to hide that. That's where I've been seeing bugs.